File: commands/Developer/npm.js
const { Command } = require("klasa");
const { MessageEmbed } = require("discord.js");
const ladybug = require("ladybug-fetch");

class NPM extends Command {
  constructor(...args) {
    super(...args, {
      description: "Search the NPM Registry for a package information",
      usage: "<package:string>",
      aliases: ["npmpackage", "npmpkg", "nodepackagemanager"],
      cooldown: 5
    });
  }

  async run(msg, [pkg]) {
    const { body } = await ladybug(`https://registry.npmjs.com/${pkg}`)
      .catch((err) => {
        if(err.status === 404) throw "No results found.";
        throw err;
      });

    const version = body.versions[body["dist-tags"].latest];

    let deps = version.dependencies ? Object.keys(version.dependencies) : null;
    let maintainers = body.maintainers.map((user) => user.name);

    if(maintainers.length > 10) {
      const len = maintainers.length - 10;
      maintainers = maintainers.slice(0, 10);
      maintainers.push(`...${len} more.`);
    }

    if(deps && deps.length > 10) {
      const len = deps.length - 10;
      deps = deps.slice(0, 10);
      deps.push(`...${len} more.`);
    }

    const embed = new MessageEmbed()
      .setColor(0xff0000)
      .setTitle(`NPM - ${pkg}`)
      .setURL(`https://npmjs.com/package/${pkg}`)
      .setAuthor(msg.author.tag, msg.author.displayAvatarURL())
      .setDescription([
        body.description || "No Description.",
        `**Version:** ${body["dist-tags"].latest}`,
        `**License:** ${body.license}`,
        `**Author:** ${body.author ? body.author.name : "Unknown"}`,
        `**Modified:** ${new Date(body.time.modified).toDateString()}`,
        `**Dependencies:** ${deps && deps.length ? deps.join(", ") : "None"}`
      ].join("\n"));

    return msg.send({ embed });
  }
}

module.exports = NPM;

File: commands/developer/npm.go
package developer

import (
  "github.com/sapphire-cord/sapphire"
  "github.com/dustin/go-humanize"
  "net/http"
  "io/ioutil"
  "encoding/json"
  "strings"
  "fmt"
  "time"
)

// Search the NPM registry for information about a package.
// Usage: <package:string>
// Aliases: npmpackage, npmpkg, nodepackagemanager
// Cooldown: 5
func NPM(ctx *sapphire.CommandContext) {
  res, err := http.Get("https://registry.npmjs.com/" + ctx.Arg(0).AsString())
  if err != nil {
    ctx.Error(err)
    return
  }

  defer res.Body.Close()

  if res.StatusCode != 200 {
    ctx.Reply("No results found.")
    return
  }

  buf, err := ioutil.ReadAll(res.Body)
  if err != nil {
    ctx.Error(err)
    return
  }

  var data struct {
    License string `json:"license"`
    Description string `json:"description"`
    Tags struct {
      Latest string `json:"string"`
    } `json:"dist-tags"`
    Versions map[string]struct {
      Dependencies map[string]interface{} `json:"dependencies"`
    } `json:"versions"`
    Maintainers []struct {
      Name string `json:"name"`
    } `json:"maintainers"`
    Author struct {
      Name string `json:"name"`
    } `json:"author"`
    Time struct {
      Modified time.Time `json:"modified"`
    } `json:"time"`
  }

  err = json.Unmarshal(buf, &data)
  if err != nil {
    ctx.Error(err)
    return
  }

  version := data.Versions[data.Tags.Latest]

  description := "No Description."

  if data.Description != "" {
    description = data.Description
  }

  author := "Unknown"

  if data.Author.Name != "" {
    author = data.Author.Name
  }

  maintainers := make([]string, 0)
  deps := make([]string, 0)

  for k, _ := range version.Dependencies {
    deps = append(deps, k)
  }

  for _, v := range data.Maintainers {
    maintainers = append(maintainers, v.Name)
  }

  if len(maintainers) > 10 {
    length := len(maintainers) - 10
    maintainers = maintainers[:10]
    maintainers = append(maintainers, fmt.Sprintf("...%d more.", length))
  }

  if len(deps) > 10 {
    length := len(deps) - 10
    deps = deps[:10]
    deps = append(deps, fmt.Sprintf("...%d more.", length))
  }

  if len(deps) == 0 {
    deps = append(deps, "None")
  }

  ctx.BuildEmbed(sapphire.NewEmbed().
    SetTitle("NPM - " + ctx.Arg(0).AsString()).
    SetColor(0xDFAC7C).
    SetURL("https://npmjs.com/package/" + ctx.Arg(0).AsString()).
    SetAuthor(ctx.Author.Username, ctx.Author.AvatarURL("256")).
    SetDescription(fmt.Sprintf("%s\n\n❯ **Version:** %s\n❯ **License:** %s\n❯ **Author:** %s\n❯ **Modified:** %s\n❯ **Dependencies:** %s\n❯ **Maintainers:** %s",
      description,
      data.Tags.Latest,
      data.License,
      author,
      humanize.Time(data.Time.Modified),
      strings.Join(deps, ", "),
      strings.Join(maintainers, ", "),
  )))
}
